lyse/__init__.py

```diff
@@ -296,6 +299,102 @@ def h5_path(self):
+
+        Examples:
+            >>> from lyse import *
+            >>> shot = Run(path)
+            >>> with shot.open('a'):
+            >>>     # shot processing that requires reads/writes
+            >>>     t, vals = shot.get_trace('my_trace')
+            >>>     results = vals**2
+            >>>     shot.save_result_array('my_result', results)
+            >>> # Shot processing that doesn't require h5 reads/writes
+            >>> print(f'Max of results is {results.max():.3f}')
+            Max of results is 5.003
+
+        Raises:
+            PermissionError: If the Run is set as read-only but a write mode is requested
+        """
+
+        # ensure no_write is respected
+        if self.no_write:
+            if mode == 'r':
+                pass
+            else:
+                msg = f'Cannot open file with a write mode; this run is read-only'
+                raise PermissionError(msg)
+            
+        # ensure file is not already opened
+        if self.__h5_file is not None:
+            msg = 'h5_file already opened'
+            raise PermissionError(msg)
+        
+        self.__h5_file = h5py.File(self.h5_path, mode)
+        try:
+            yield None
+        finally:
+            self.__h5_file.close()
+            self.__h5_file = None
+            
+    def open_file(mode):
+        """Wrapper for lyse functions to allow using previously opened file with context manager.
+        
+        If multiple read/write operations happen on a Run in a single shot,
+        opening the h5file once via the context manager `open`
+        can speed up the analysis execution time by limiting the number of times
+        the file must be opened/closed.
+
+        Args:
+            mode (str): which mode to open the h5 file with.
+                Must be 'r', 'a', 'r+', 'w', 'w-', or 'x'.
+                Lyse typically only uses 'r' and 'a'.
+
+        Raises:
+            PermissionError: If the Run is set as read-only but a write mode is requested
+        """
+        def decorator_open_file(func):
+            @functools.wraps(func)
+            def wrapper(self, *args, **kwargs):
+
+                if self.no_write:
+                    if mode == 'r':
+                        pass
+                    else:
+                        msg = f'Cannot perform operation {func.__name__:s}; this run is read-only'
+                        raise PermissionError(msg)
+
+                if self.__h5_file is None:
+                    with self.open(mode):
+                        return func(self, *args, **kwargs)
+                else:
+                    if (self.__h5_file.mode == 'r') and (mode != 'r'):
+                        msg = (f"Cannot perform operation {func.__name__:s}; "
+                            + f"requested mode {mode:s} not compatible with "
+                            + f"h5_file's mode {self.__h5_file.mode:s}")
+                        raise PermissionError(msg)
```

Comment:
I think some additional documentation around this (either comment here or in docstring) would be useful. It took me a while to work out that passing in `r+` as the mode (e.g. see use of decorator further down file) wasn’t incompatible with someone manually opening the file with `a` for example. I think this is something that will be easily forgotten unless we document it now. The only incompatible case is the one for this error message, where if the file is open in read only, then any decorator use that wants write access needs to throw an error.

That could be an argument for not actually having the argument to the decorator be `mode`, but a `Boolean` that indicates whether we need it to be open in a writable mode or not. That boolean would then toggle between `r` or `a` in the call to `self.open()` a few lines above.

Reply:
Which brings me to another point, it seems like we’ve switched the use of `a` to `r+` which would be a backwards incompatible change?

Reply:
Yeah, this is due to the quirks of how h5py File objects work. While there are a number of mode options for [opening the file](https://docs.h5py.org/en/stable/high/file.html#opening-creating-files), the file itself can only be in [two modes](https://docs.h5py.org/en/stable/high/file.html#h5py.File.mode) (`'r'` and `'r+'`). All the other options for opening the file deal exclusively with how to handle creating a new file. I'll update the docstring to make this clear as it isn't super intuitive at first glance.

This relates to my thought process for changing most of the calls from `a` to `r+`. The only difference between these two is that `a` will create the file if it doesn't exist already. It is my understanding that almost all lyse functions should never create a new file on their own: that job is handled by runmanager or the `Sequence` constructor. For example, if `save_result` is creating a new file, I'd argue that is a bug that should raise an error to be fixed instead of failing silently as it would now. Assuming that reasoning is valid, I see there are a few docstrings that should be updated to reflect that. I've gone ahead and changed them. If my reasoning is wrong, I can change things back.

I decided to keep mode strings (instead of a bool) to match the h5py documentation. Felt like it would be more clear what was happening if we didn't add another layer of abstraction over top.

Reply:
A reason for `a` might be if you're using `Run` as a multishot analysis storage system. I suppose that's probably a niche use case and we can just require you create the file first in that case.

Thoughts @chrisjbillington ?

Reply:
It is niche, and we shouldn't worry much about niche things.

Though for what it's worth, I also think worrying about someone passing in a non-existent file unintentionally is fairly niche. So I don't think it really matters.

On balance though, I lean toward discouraging using `Run` for multi-shot storage, because that means we have to continue to commit to ensuring it works with arbitrary HDF5 files, whereas it would be nice to be able to assume the HDF5 file is an actual shot file. It currently works with arbitrary files, but minimising future commitments is good.

And if someone wants to use HDF5 to store data, they may do so with `h5py` directly - our wrapper is not adding much and is actually pretty limiting. I would encourage using raw `h5py` instead for this, so I have no problem making `Run` less convenient to use for this, it is unlikely to be the best tool for the job anyway.

So +1 for `r+` I think.

Reply:
Cool. I'll keep as-is then.

Also, I'm pretty sure `Run` will currently fail if called on a non-existing file thanks to `_create_group_if_not_exists`, which already uses `r+`, so hopefully this won't break anyone's workflow.

https://github.com/labscript-suite/lyse/blob/82db03237f8667ba65859c1fb9e647a123bfb98e/lyse/__init__.py#L331-L341